### PR TITLE
Add passthrough delimiter for argument parser

### PR DESCRIPTION
**Pull request by arezaii**
This PR adds passthrough arguments to the argument parser. 

Scope of work defined in Cray/chapel-private#2304, this work satisfies
the main goals of that issue which are to:

1. Allow the developer to specify a delimiter meaning:
 "consume everything past this point as-is"

Relates to broader implementation task in Cray/chapel-private#2192

This PR also incorporates refactoring some class and variable names.

Tests added for passthrough and mixing positional arguments, subcommands, options, 
and flags with passthrough arguments. 

Example updated to include passthrough, and several new end-to-end tests added
for both good and bad command-line entries.

TESTING:

After running `make cleanall` from `$CHPL_HOME`:

- [x] Can `make`
- [x] Can `make docs`
- [x] Can `make mason`
- [x] Can `make check`
- [x] Passing tests from `util/start_test test/mason/mason-argparse`

Reviewed by @mppf 

Signed-off-by: Ahmad Rezaii ahmad.rezaii@hpe.com

### CODE REVIEW

**Comment by mppf** [comment]
It looks like this is using `++` for the delimiter (vs the IMO more common `--`). I understand there are some issues with `--` being consumed by Chapel's config var argument parsing. But is the choice to use `++` a near term one or something that was discussed? Did we consider other ways of handling the `--` from Chapel's argument parsing? For example, we might use an approach where, if a program wants to support `--`, it gets all of the arguments to `main` and then calls Chapel's argument parsing somehow. (See https://github.com/Cray/chapel-private/issues/2289#issuecomment-883437321 ). Or, maybe, Chapel's argument parsing can simply leave the `--` there when it is present.

**Comment by arezaii** [comment]
> It looks like this is using `++` for the delimiter (vs the IMO more common `--`).

Correct, the `--` is never seen unless it is placed in the command twice, due to the runtime consuming the first instance it finds.

> But is the choice to use ++ a near term one or something that was discussed?

I used `++` as a default to be a workaround to this problem in the short term. It is still a parameter though, and can be set to whatever the user desires. However, just setting it to `--` won't change the runtime's behavior and will likely lead to some confused users.

> Did we consider other ways of handling the -- from Chapel's argument parsing?

I will have to do some investigating to see how we could get the runtime to optionally not consume the `--`. I considered that to be out of scope for this task rather than a dependency, as it's something that could be changed later without much consequence here, aside from updating some tests and the default value. I will add a spike to the next sprint to look into it.

File: tools/mason/MasonArgParse.chpl [hunk 1]
```diff
@@ -766,17 +757,30 @@ module MasonArgParse {
...
                                     required=required,
                                     defaultValue=myDefault
                                     );
-      return _addAction(action);
+      return _addHandler(handler);
     }
   }
 
+  // helper to prepare numArgs ranges for use
+  proc _prepareRange(rIn: range(?)) :range throws {
```
Comment (mppf) [inline tools/mason/MasonArgParse.chpl:772]: Let's put a space after the colon in the return type. (I'm ok with the space before it being present, or not).

File: test/mason/mason-argparse/MasonArgParseExample.good [hunk 2]
```diff
@@ -38,3 +44,9 @@ one
 two
 subBoolFlag: true
 subcommand positional value: bValue
+++
```
Comment (mppf) [inline test/mason/mason-argparse/MasonArgParseExample.good:47]: It seems surprising to me that `passedThrough.values()` includes the `++`. At the least this is worth some docs comments in `addPassThrough` but you might consider taking it out for pass throughs.

I guess in the case of a subcommand it corresponds to `argv[0]` being the name of the program? (I.e. it is now the name of the subcommand)?

**Review by mppf — APPROVED** [review]
Sounds good. I think it'd be good to study if we can use `--` instead of `++` but as we discussed that will be future work.

File: test/mason/mason-argparse/MasonArgParseExample.good (hunk 2 above)
Comment (arezaii) [reply test/mason/mason-argparse/MasonArgParseExample.good:47]: > I guess in the case of a subcommand it corresponds to argv[0] being the name of the program? (I.e. it is now the name of the subcommand)?

Yes, exactly that. I thought that would be a consistent way to approach it.


> It seems surprising to me that passedThrough.values() includes the ++.

Yeah, I was on the fence about that one too. Your comment pushed me to create a new sub-class and treat the delimiter for passthroughs differently than the subcommand.

File: tools/mason/MasonArgParse.chpl [hunk 3]
```diff
@@ -128,18 +126,17 @@ module MasonArgParse {
...
       debugTrace("Searching positions from: " + pos:string + " to "
                  + endPos:string);
 
       if args[pos] == this._name {
-        myArg._values.append(args[pos]);
+        myArg._values.extend(args[pos..]);
         debugTrace("matched cmd: " + args[pos] + " at pos: " + pos:string);
-        rest.extend(args[next..]);
-        return next;
+        return argsD.high + 1;
```
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:139]: updated to be more consistent

**Pull request merged**